### ISSUE TITLE
Additional unicode character issue

### ISSUE THREAD

**Issue by ReToCode** [sematext/sematext-agent-docker#26]
Hi guys,

I work at SBB with @szakasz. We have an unicode character that gets added to the logs some times. Example see here (https://apps.sematext.com/ui/logs/10050/0?endDate=1490856600000&query&queryFilters=%5B%7B%22f%22%3A%22kubernetes.namespace.raw%22%2C%22q%22%3A%5B%22zvs-orchestrierung-zvs-inte%22%5D%7D%5D&startDate=1490770200000). See the message from "08:04:00.000" in field severity: IN�FO.

I already debugged this issue up the dependency chain up to dockerode itself. I think it has something to do with the demultiplexing of the stream (See https://github.com/docker/docker/issues/7375). I also tried with the docker-modem parser of dockerode, does not help (https://github.com/SchweizerischeBundesbahnen/sematext-file-agent/blob/master/lib/loghose/loghose.js#L66).

I also noticed, that sometimes the actual full length of the chunk is smaller than the value that's specified in the 4-bytes of the header. In the end I think it mixes up logs & header somehow and reads a part of the header as logs which results in those characters. Any ideas? Heard anything like it? We're using docker 1.10.3.

**Comment by megastef** [comment]
Thanks for reporting. I think we had such an issue in the past https://github.com/mcollina/docker-loghose/issues/4 after the update of docker-loghose we did not see the problem anymore. Have you tried to update to Docker Engine to 1.12 or higher?

**Comment by szakasz** [comment]
A bit better link is this: https://apps.sematext.com/ui/logs/10050/0?endDate=1490854800000&query&queryFilters=%5B%7B%22f%22%3A%22kubernetes.namespace.raw%22%2C%22q%22%3A%5B%22zvs-orchestrierung-zvs-inte%22%5D%7D%5D&startDate=1490847600000

Here both the first log entry has a problem in the severity field and also the second one in the message at "dauerAntw1�ort".

**Comment by megastef** [comment]
Thx, we are looking into it. Please make sure that you run SDA v1.31.23 - the temporary release 1.31.22 had a bug, creating invalid logs. 
When did the problem with invalid characters start?

**Comment by ReToCode** [comment]
We're currently on 1.31.20. So this should not be the case. We are running OpenShift as orchestrator so we are not flexible with docker versions. Thanks for the links, I'll look into them as well. 
Hard to tell when it started. We noticed it in the new UI for the first time. I just tried to search the logs for this characters, but the search engine seems not to like that :D

**Comment by megastef** [comment]
We see the issue only with one image in last 24 hours. Maybe the one has different TTY setting? 
Query for `severity:in AND severity:fo` - the \u0000 breaks the word so Info gets tokenized to `In` and `fo`.  You can see the relevant images in by expanding image_name in the filters and set it to "current query".

https://apps.sematext.com/ui/logs/APPID/0?endDate=1490873400000&query=severity%3Ain+AND+severity%3Afo&queryFilters=%5B%5D&startDate=1490787000000

**Comment by szakasz** [comment]
The special characters come at different locations (see my comment from yesterday), so once it gets caught at IN...FO and at other times (and that's more often) it's in the message. The link I sent yesterday contains both cases.

**Comment by megastef** [comment]
Did you try 1.31.25? We moved to latest docker-loghose (before we used a patched version). If the problems remains we have to work on a fix for docker-loghose.

**Comment by ReToCode** [comment]
We are now on 1.31.25 for 24h. @szakasz How can we check if the problem remains? I don't see any logs based on the change with the default=false :)

**Comment by szakasz** [comment]
I don't think there's a generic search query in case of this special chars. :(
If we look for "severity:in AND severity:fo" then such cases happen about once a day although with high deviation. Now this hasn't happened for 3 days, however considering the big drop in the log amount since monday, it's not sure that the issue is solved. If we don't see it anymore for another 10 days at least, then we can claim, it's solved. :)

**Comment by megastef** [comment]
I think the problem was caused by very large logs (like large XML requests/responses). Such logs get divided in multiple chunks with a leading header in the Docker log stream.  So the best test case would be to enable logging for containers, which had such problematic logs the past. As far I know this is solved in latest docker-loghose.

**Comment by ReToCode** [comment]
We discussed the large xmls as reason as well. Which version of docker-loghose do you mean? This diff? https://github.com/mcollina/docker-loghose/compare/master...megastef:master.?

I see the problem also using dockerode itself: 

```
// loghose.js
container.attach({stream: true, stdout: true, stderr: true}, function (err, stream) {
                if (err) {
                    // no stream availaable
                    return
                }
                let demuxedStream = new nodeStream.PassThrough();
                container.modem.demuxStream(stream, demuxedStream, demuxedStream);

                streams[data.id] = stream;

                demuxedStream.pipe(through.obj(function (c,enc,cb) {
                    this.push({
                        id: data.id.slice(0, 12),
                        image: data.image,
                        name: data.name,
                        time: Date.now(),
                        line: c.toString()
                    });
                    cb();
                })).pipe(result, {end: false}).on('error', function () {
                    /* continue, let's wait for docker restart */
                });
            })

```

I used dockerode directly with their demux and see the same issues. I'm trying to find a logline that causes the error every time.

**Comment by megastef** [comment]
In docker-loghose a parser is used to parse headers: 
https://github.com/mcollina/docker-loghose/blob/master/lib/loghose.js#L65
https://github.com/mcollina/docker-loghose/blob/master/lib/parser.js

**Comment by ReToCode** [comment]
Yes I saw this one. I also checked for the tty settings. No container has tty true at our environment. So the problem happens in the parser somewhere or the docker api itself. 

As you said above, the problem seems to be if a very long log entry gets divided in multiple chunks, then the header gets parsed as well - but I was not able to see why so far.

**Comment by ReToCode** [comment]
I think I found the problem in the parser. We indeed have sometimes very long lines from a lot of attached containers then the following happens:

The parse function gets called with a fragment that is not a header:
```
 function parse (chunk, enc, cb) {
        var buffer = bl(chunk)
        var full = buffer.length;

        var h = buffer.get(0) + '-' + buffer.get(1) + '-' + buffer.get(2) + '-' + buffer.get(3) + '-' + buffer.get(4) + '-' + buffer.get(5) + '-'+ buffer.get(6) + '-' + buffer.get(7);

        if (buffer.get(0) > 3) {
                console.log('wrong', buffer.get(0));
        }
```
We here get a lot of following values for 'h':
1-0-0-0-0-0-128-0 // correct
68-55-100-75-80-103-108-47 // definitely not a header > this results in a wrong payloadLength
116-62-10-9-9-60-47-116 // and so on

I tried the solution with the docker-modem-parser:
```
Modem.prototype.demuxStream = function(stream, stdout, stderr) {
  var header = null;

  stream.on('readable', function() {
    header = header || stream.read(8);
    while (header !== null) {
      var type = header.readUInt8(0);
      var payload = stream.read(header.readUInt32BE(4));
      if (payload === null) break;
      if (type == 2) {
        stderr.write(payload);
      } else {
        stdout.write(payload);
      }
      header = stream.read(8);
    }
  });
};
```
And call it like this:
```
let demuxedStream = new nodeStream.PassThrough();
                container.modem.demuxStream(stream, demuxedStream, demuxedStream);

                streams[data.id] = stream;

                demuxedStream.pipe(through.obj(function (c,enc,cb) {
                    this.push({
                        id: data.id.slice(0, 12),
                        image: data.image,
                        name: data.name,
                        time: Date.now(),
                        line: c.toString()
                    });
                    cb();
                })).pipe(result, {end: false}).on('error', function () {
                    /* continue, let's wait for docker restart */
                });
```
This seems to solve it.

**Comment by megastef** [comment]
@ReToCode could provide a PR for docker-loghose or docker-modem?

**Comment by ReToCode** [comment]
Yes. I'll see how it goes the next week, then create one.

**Comment by megastef** [comment]
@ReToCode Please try sematext/sematext-agent-docker:dev image. It includes the long-line-fix from docker-loghose https://github.com/mcollina/docker-loghose/issues/4#issuecomment-296334766 

We generated long log lines (50k) and could not find any invalid character.

**Comment by ReToCode** [comment]
Tested it today. Seems to work fine, tested for a few hours!

**Comment by megastef** [comment]
FYI. sematext/sematext-agent-docker 1.31.27 is already using the patch https://github.com/sematext/sematext-agent-docker/blob/1.31.27/package.json#L54
We will update it to the final version as soon new docker-loghose version is ready.

**Comment by megastef** [comment]
Published today docker-loghose 1.5.1 - next SDA release will use official docker-loghose again
